source/compiler/qsc_eval/src/backend.rs

```diff
@@ -665,190 +665,237 @@ fn unwrap_matrix_as_array2(matrix: Value, qubits: &[usize]) -> Array2<Complex<f6
+        self.backend.t(q);
+        if let Some(tracer) = &mut self.tracer {
+            tracer.t(q);
+        }
     }
 
     fn x(&mut self, q: usize) {
-        self.chained.x(q);
-        self.main.x(q);
+        self.backend.x(q);
+        if let Some(tracer) = &mut self.tracer {
+            tracer.x(q);
+        }
     }
 
     fn y(&mut self, q: usize) {
-        self.chained.y(q);
-        self.main.y(q);
+        self.backend.y(q);
+        if let Some(tracer) = &mut self.tracer {
+            tracer.y(q);
+        }
     }
 
     fn z(&mut self, q: usize) {
-        self.chained.z(q);
-        self.main.z(q);
+        self.backend.z(q);
+        if let Some(tracer) = &mut self.tracer {
+            tracer.z(q);
+        }
     }
 
     fn qubit_allocate(&mut self) -> usize {
-        // Warning: we use the qubit id allocated by the
-        // main backend, even for later calls into the chained
-        // backend. This is not an issue today, but could
-        // become an issue if the qubit ids differ between
-        // the two backends.
-        let _ = self.chained.qubit_allocate();
-        self.main.qubit_allocate()
+        let q = self.backend.qubit_allocate();
+        if let Some(tracer) = &mut self.tracer {
+            tracer.qubit_allocate(); // ignored
+        }
+        q
     }
 
     fn qubit_release(&mut self, q: usize) -> bool {
-        let _ = self.chained.qubit_release(q);
-        self.main.qubit_release(q)
+        let b = self.backend.qubit_release(q);
+        if let Some(tracer) = &mut self.tracer {
+            tracer.qubit_release(q);
+        }
+        b
     }
 
     fn qubit_swap_id(&mut self, q0: usize, q1: usize) {
-        self.chained.qubit_swap_id(q0, q1);
-        self.main.qubit_swap_id(q0, q1);
+        self.backend.qubit_swap_id(q0, q1);
+        if let Some(tracer) = &mut self.tracer {
+            tracer.qubit_swap_id(q0, q1);
+        }
     }
 
     fn capture_quantum_state(
         &mut self,
     ) -> (Vec<(num_bigint::BigUint, num_complex::Complex<f64>)>, usize) {
-        let _ = self.chained.capture_quantum_state();
-        self.main.capture_quantum_state()
+        self.backend.capture_quantum_state()
     }
 
     fn qubit_is_zero(&mut self, q: usize) -> bool {
-        let _ = self.chained.qubit_is_zero(q);
-        self.main.qubit_is_zero(q)
+        self.backend.qubit_is_zero(q)
     }
 
     fn custom_intrinsic(&mut self, name: &str, arg: Value) -> Option<Result<Value, String>> {
-        let _ = self.chained.custom_intrinsic(name, arg.clone());
-        self.main.custom_intrinsic(name, arg)
+        let res = self.backend.custom_intrinsic(name, arg.clone());
+        if let Some(tracer) = &mut self.tracer {
+            tracer.custom_intrinsic(name, arg);
+        }
+        res
     }
 
     fn set_seed(&mut self, seed: Option<u64>) {
-        self.chained.set_seed(seed);
-        self.main.set_seed(seed);
+        self.backend.set_seed(seed);
     }
```

Comment:
Are we omitting seed since the tracer won't use it?

Reply:
Yes, that was my intention. Much like we don't use the return value from the tracer in some of the methods (like `m`), this is one where the behavior of the two "backends" aren't exactly parallel.

This method will be dropped altogether in my next PR when `Tracer` evolves away from `Backend`.